File: SabreCG_PY/Lof.py
import util as ut
from typing import List
from Leg import Leg
from Aircraft import Aircraft
from OperLeg import OperLeg
from Station import Station

class Lof:
    _count = 0

    def __init__(self):
        self._aircraft = None
        self._cost = 0
        self._id = _count
        _count += 1

    def pushLeg(self, leg: OperLeg) -> None:
        self._legList.append(leg)
        if leg.getLeg().isMaint():
            self._maintList.append(leg)


    def popLeg(self) -> None:
        self._legList.pop()
    
    def getSize(self) -> int:
        return len(self._legList)
    
    def setLegList(self, legList: List[OperLeg]) -> None:
        self._legList = legList

    def getLegList(self) -> List[OperLeg]:
        return self._legList
    
    def setAircraft(self, aircraft: Aircraft) -> None:
        self._aircraft = aircraft

    def getAircraft(self) -> Aircraft:
        return self._aircraft
    
    def computeLofCost(self) -> None:
        self._cost = 0
        for _leg in self._legList:
            if not _leg.isMaint():
                self._cost += (_leg.getOpDepTime() - _leg.getScheDepTime())/60.0 * ut.util.w_fltDelay
            if _leg.getScheAircraft() != self._aircraft:
                self._cost += ut.util.w_fltSwap

    def containMaint(self) -> bool:
        return len(self._maintList) > 0

    def setCost(self, cost: float) -> None:
        self._cost = cost

    def getCost(self) -> float:
        return self._cost

    def getLastLeg(self) -> OperLeg:
        return self._legList[-1]
    
    def print(self) -> None:
        print("************ LOF ************")
        print("Lof ID is ")
        if self._aircraft != None:
            self._aircraft.print()
        else:
            print("Tail is NULL")
        print("Leg List Size is %d cost %d" % (len(self._legList), self._cost))
        print()
        for _leg in self._legList:
            _leg.print()
        print("*****************************")

    def getDepStation(self) -> Station:
        return self._legList[0].getLeg().getDepStation()
    
    def getArrStation(self) -> Station:
        return self._legList[-1].getLeg().getArrStation()
    
    def getOperDepTime(self) -> float:
        return self._legList[0].getOpDepTime()

    def getOperArrTime(self) -> float:
        if self._legList[-1].getLeg().isMaint():
            return self._legList[-1].getOrArrTime()
        return self._legList[-1].getOpArrTime()

    def getDepTime(self) -> float:
        return self._legList[0].getPrintDepTime()

    def computeReducedCost(self) -> None:
        self._reducedCost = 0
        sumLegDual = 0
        for _leg in self._legList:
            sumLegDual += _leg.getLeg().getDual()
        self._reducedCost = self._cost - sumLegDual - self._aircraft.getDual()

    def getReducedCost(self) -> float:
        return self._reducedCost

    def setId(self, id: int) -> None:
        self._id = id

    def getId(self) -> int:
        return self._id
    
    def compareDepTimeKey(self) -> float:
        return self.getOperDepTime()

File: SabreCG_multipara/SabreCG/Lof.cpp
#include "Lof.h"

Lof::Lof()
{
	_aircraft = NULL;
	_cost = 0;
	//* _purity = 0;
	//* _id = 0;
	_id = _count;			//*
	_count++;				//*
}

void Lof::pushLeg(OperLeg * leg)
{
	_legList.push_back(leg);

	if( leg->getLeg()->isMaint() )
	{
		_maintList.push_back(leg);
	}
}

int Lof::_count = 0;		//*

/*
//* duplicate Lof and assign aircraft
Lof * Lof::clone()
{
	Lof * lof = new Lof();

	for(int i = 0; i < _legList.size(); i++)
	{
		OperLeg * newLeg = _legList[i]->clone();     // ScheduledLeg + depTime + arrTime + OpeAircraft
		lof->pushLeg(newLeg);
	}

	//	lof->setLegList(_legList);
	lof->setAircraft(_aircraft);
	lof->setCost(_cost);
	lof->setPurity(_purity);
	return lof;
}
*/

Lof::~Lof()
{
	for(int i =0; i < _legList.size(); i++)
	{
		delete _legList[i];
	}
}

void Lof::print()
{
	cout <<"************ LOF ************" << endl;
	cout<< "Lof ID is " << _id <<endl;
	if ( _aircraft != NULL)
	{
		_aircraft->print();
	}
	else
	{
		cout << "Tail is NULL" << endl; //*
	}
	cout <<"Leg List Size is " << _legList.size() << " cost " << _cost <<  endl << endl;
	//	cout <<"Leg List Size is " << _legList.size() << " purity " << _purity <<  endl << endl;

	for(int i = 0; i < _legList.size(); i++)
	{
		_legList[i]->print();
	}

/*	if (_depNode!=NULL)
	{
		cout<<"ID of depNode "<<_depNode->getId()<<endl;
			cout<<"departure station is "<<_depNode->getStation()->getName()<<endl;
		if (_depNode->getAirctaft() != NULL)
		{
		cout<<"aircraft is "<<_depNode->getAirctaft()->getTail()<<endl;
		}
		time_t depTime = _depNode->getTime() - TIMEDIFF;
		cout<<"Node time is "<< ctime(&depTime)<<endl;
	/*	if (_depNode->getNextNode()!=NULL)
		{
		cout<<"ID of next node "<<_depNode->getNextNode()->getId()<<endl;
		cout<<"station of next node is "<<_depNode->getNextNode()->getStation()->getName()<<endl;
		cout<<"aircraft is "<<_depNode->getNextNode()->getAirctaft()->getTail()<<endl;
		time_t Time = _depNode->getNextNode()->getTime() - TIMEDIFF;
		cout<<"departure time is "<< ctime(&Time)<<endl;
		}*/
//	}
/*	if (_arrNode!=NULL)
	{
		cout<<"ID of arrNode "<<_arrNode->getId()<<endl;
			cout<<"arrival station is "<<_arrNode->getStation()->getName()<<endl;
		if (_arrNode->getAirctaft() != NULL)
		{
		cout<<"aircraft is "<< _arrNode->getAirctaft()->getTail()<<endl;
		}
		time_t arrTime = _arrNode->getTime() - TIMEDIFF;
		cout<<"Node time is "<< ctime(&arrTime)<<endl;
	/*	if (_arrNode->getNextNode()!=NULL)
		{
		cout<<"ID of next node "<<_arrNode->getNextNode()->getId()<<endl;
		cout<<"station of next node is "<<_arrNode->getNextNode()->getStation()->getName()<<endl;
		cout<<"aircraft is "<<_arrNode->getNextNode()->getAirctaft()->getTail()<<endl;
		time_t Time = _arrNode->getNextNode()->getTime() - TIMEDIFF;
		cout<<"arrival time is "<< ctime(&Time)<<endl;
		}*/

//	}
	cout <<"*****************************"  << endl;

}

/*
void Lof::compPurity()
{
	vector<Aircraft *> usedAircraftList;
	for(int i = 0; i < _legList.size(); i++)
	{
		bool found = false;
		for(int j = 0; j < usedAircraftList.size(); j++)
		{
			if ( _legList[i]->getLeg()->getAircraft() == usedAircraftList[j])
			{
				found = true;
				break;
			}
		}

		if ( found == false)
		{
			usedAircraftList.push_back( _legList[i]->getLeg()->getAircraft());
		}
	}
	_purity = usedAircraftList.size();
}
*/

/*
bool Lof::compatible(Aircraft * aircraft)
*/

/*
	bool rescheduleTime(time_t t);                         /// ????????????????????????????????????? 都是处理sequential delay的?
	bool rescheduleTime();
	bool rescheduleTimeWithMaint();
	bool rescheduleTimeWithoutMaint();
	bool rescheduleTimeWithMaint(time_t prevEndTime);
	bool rescheduleTimeWithoutMaint(time_t prevEndTime);
	void rescheduleTimeWithoutMaintAirportClose();
	void rescheduleTimeWithMaintAirportClose();
	*/

/* compute the cost of the lof = delay + swap */
/*
void Lof::compCostWithoutPopulate()
{
	_cost = 0;
	for(int i = 0; i < _legList.size(); i++)
	{
		if ( !_legList[i]->getLeg()->isMaint() )
		{
			_cost = _cost + (_legList[i]->getOpDepTime() - _legList[i]->getScheDepTime())/60 * Util::w_fltDelay;
		}
	}
	compPurity();
	_cost = _cost + _purity * Util::w_fltSwap;
}
*/

void Lof::computeLofCost()
{
	_cost = 0;
	for(int i = 0; i < _legList.size(); i++)
	{
		if ( !_legList[i]->getLeg()->isMaint() )
		{
			_cost = _cost + (_legList[i]->getOpDepTime() - _legList[i]->getScheDepTime())/60.0 * Util::w_fltDelay; /* 注意这里除以60 */
			//_cost = _cost + (_legList[i]->getOpDepTime() - _legList[i]->getScheDepTime()) * Util::w_fltDelay; // 小case debug用
		}

		if ( _legList[i]->getScheAircraft() != _aircraft)
		{
			_cost = _cost + Util::w_fltSwap;
		}
	}
}

time_t Lof::getOperArrTime()
{ // with turn time added, can depart immediately...
	if ( _legList.back()->getLeg()->isMaint() )
	{
		return _legList.back()->getOpArrTime();
	}
	//* return _legList.back()->getOpArrTime() + Util::turnTime;
	return _legList.back()->getOpArrTime();
}

time_t Lof::getOperDepTime()
{ // with turn time deducted if maint, can depart immediately
	/*	if ( _legList.front()->getLeg()->isMaint())
	{
	return _legList.front()->getOpDepTime() + Util::turnTime;
	}
	*/
	return _legList.front()->getOpDepTime();
}

time_t Lof::getDepTime()
{
	return _legList.front()->getPrintDepTime();
}

bool Lof::compareDepTime(Lof * a, Lof *b)
{
	if(a->getOperDepTime() < b->getOperDepTime() )
	{
		return true;
	}
	return false;
}

bool Lof::compareByAircraft(Lof* a, Lof* b)
{
	if (a->getAircraft()->getId() < b->getAircraft()->getId())	// 比较aircraft
	{
		return true;
	}

	if (a->getAircraft()->getId() == b->getAircraft()->getId())	// aircraft相同，比较lof id
	{
		return a->getId() < b->getId();
	}

	return false;

}

void Lof::computeReducedCost()
{
	_reducedCost = 0;
	double sumLegDual = 0;		//*

	/* flow balance constraint */
	//* _reducedCost = _depNode->getDual() - _arrNode->getDual();     ///********注意depNode是正，arrNode是负 /// flow balance constraint的dual是什么时候被set的？

	/* cover constraint */

	//*cout << "_legList.size() is " <<  _legList.size() << endl;

	for (int i = 0; i < _legList.size(); i++)
	{
		//* _reducedCost += _legList[i]->getLeg()->getDual();
		sumLegDual += _legList[i]->getLeg()->getDual();
	}

	//* _reducedCost = _reducedCost - getCost();

	_reducedCost = _cost - sumLegDual - _aircraft->getDual();

}